src/includes/dinit-iostream.h

```diff
@@ -0,0 +1,741 @@
+    ssize_t flush();
+
+    // write() functions, Used to write many types.
+    //
+    // All calls to write() will be buffered and not necessarily written to the file descriptor except
+    // in these situations:
+    // 1. The size of the passed message is larger than the buffer size.
+    // 2. The size of the passed message is larger than the current buffer free space.
+    // 3. "flush()" (or its exception-free variant: flush_nx()) is called.
+    // 4. A "flush" object is passed to write() functions (equivalent to calling flush()).
+    // 5. An "endl" object is passed to write() functions (equivalent to writing \n and
+    //    calling flush()).
+    // 6. The object is destructed (which calls flush_nx()).
+    //
+    // Multiple arguments can be passed in a single call as well (due to the templated overload), e.g:
+    //
+    //      output_a.write("This is an example message.\n");
+    //      output_b.write("2 + 2 equals: ", 2 + 2, endl);
+    //
+    // Returns: true or false on failure cases or null/zero-length string.
+    //
+    // Note: write() does return false on partial write, Use write_buf() and check its return, if you
+    // want to handle partial write gracefully.
+    //
+    // Non _nx variants may throw:
+    // dio::iostream_internal_err on Buffer failure (buffer did not allocated at construction time)
+    // dio::iostream_system_err on system I/O failure
+    bool write_nx(const char *msg) noexcept;
+    bool write(const char *msg);
+
+    bool write_nx(char msg) noexcept;
+    bool write(char msg);
+
+    bool write_nx(const std::string &msg) noexcept;
+    bool write(const std::string &msg);
+
+    bool write_nx(const int num) noexcept;
+    bool write(const int num);
+
+    bool write_nx(const unsigned num) noexcept;
+    bool write(const unsigned num);
+
+    bool write_nx(const long num) noexcept;
+    bool write(const long num);
+
+    bool write_nx(const unsigned long num) noexcept;
+    bool write(const unsigned long num);
+
+    bool write_nx(const endline &) noexcept;
+    bool write(const endline &);
+
+    bool write_nx(const flushbuf &) noexcept;
+    bool write(const flushbuf &);
+
+    template <typename A, typename B, typename ...C> inline bool write_nx(const A &a, const B &b, const C & ...c)
+            noexcept
+    {
+        if (!good()) return false;
+        if (!write_nx(a)) return false;
+        if (!write_nx(b, c...)) return false;
+        return true;
+    }
+
+    template <typename A, typename B, typename ...C> inline bool write(const A &a, const B &b, const C & ...c)
+    {
+        if (!good()) return false;
+        if (!write(a)) return false;
+        if (!write(b, c...)) return false;
+        return true;
+    }
+
+    // write_buf() functions, Used to write char* or std::string.
+    //
+    // All calls to write_buf() will be buffered and not necessarily written to the file descriptor
+    // except in situations that are mentioned about write() above.
```

Comment:
Currently an error condition will be set if a partial write occurred (see `ostream::put`). It would be better if (in `write_buf` case, but not for `write`) partial writes just return without trying to write the remainder and leave the error state clear, since this function is fine with partial writes, but it would be ok to leave it as is as long as you document it properly.

Reply:
Will be documented